core/src/main/java/com/radixdlt/messaging/core/MessageDispatcher.java

```diff
@@ -137,26 +143,82 @@ CompletableFuture<Result<Unit, Cause>> send(final OutboundMessageEvent outboundM
         .findOrCreateChannel(outboundMessage.receiver())
         .thenApply(channel -> send(channel, bytes))
         .thenApply(this::updateStatistics)
-        .exceptionally(t -> completionException(t, receiver, message));
+        .thenApply(
+            result -> {
+              result.onError(err -> logSendError(message, receiver, err.toString()));
+              return result;
+            })
+        .exceptionally(
+            t -> {
+              final var cause = t.getCause() != null ? t.getCause().getMessage() : t.getMessage();
+              logSendError(message, receiver, cause);
+              return IO_ERROR.result();
+            });
+  }
+
+  private void logSendError(Message message, NodeId receiver, String cause) {
+    final RateLimiter rateLimiter;
+    final AtomicInteger countSinceLastLog;
+    if (sendErrorLogRateLimitersByReceiver.contains(receiver)) {
+      final var curr = sendErrorLogRateLimitersByReceiver.get(receiver).orElseThrow();
+      rateLimiter = curr.getFirst();
+      countSinceLastLog = curr.getSecond();
+    } else {
+      rateLimiter = RateLimiter.create(0.05);
+      countSinceLastLog = new AtomicInteger(0);
+      sendErrorLogRateLimitersByReceiver.put(receiver, Pair.of(rateLimiter, countSinceLastLog));
```

Comment:
BTW is there a reason we have our own `LRUCache` impl?
(e.g. do we want a very strict [deterministic, predictable] "LRU" behavior?)

Or do we rather want a practically-good eviction policy, tuned for hit-rate?
(if so, then we should probably go for guava/caffeine I think?)

Reply:
I wouldn't really call it "our own impl", it's just a thin helper/wrapper around `LinkedHashMap`, no special requirements there, we could as well use Guava.